File: 0460-lfu-cache/0460-lfu-cache.java
class LFUCache {
    class Node{
        int key,value,freq;
        Node prev,next;
        Node(int key,int value){
            this.key = key;
            this.value = value;
            this.freq = 0;
        }
    }
    class DLinkedList{
        int freq;
        Node head,tail;
        DLinkedList(int freq){
            this.freq =  freq;
            head = new Node(0,0);
            tail = new Node(0,0);
            head.next = tail;
            tail.prev = head;
        }
        void add(Node node){
            Node t = head.next;
            head.next = node;
            node.prev = head;
            node.next = t;
            t.prev = node;
        }
        boolean isEmpty(){
            return head.next==tail;
        }
        void delete(Node node){
            node.next.prev = node.prev;
            node.prev.next = node.next;
        }
        
        Node pop(){
            if (isEmpty()) return null;
            Node node = tail.prev;
            delete(node);
            return node;
        }
        
        
    }
    
    int capacity,size,minFreq;
    HashMap<Integer,Node> keyMap;
    HashMap<Integer,DLinkedList> freqMap;
    
    public LFUCache(int capacity) {
        this.capacity = capacity;
        this.size = minFreq = 0;
        keyMap = new HashMap<>();
        freqMap = new HashMap<>();
        freqMap.put(0,new DLinkedList(0));
    }
    
    private int _update(Node node){
        int freq = node.freq;
        freqMap.get(freq).delete(node);
        node.freq ++;
        freqMap.computeIfAbsent(node.freq,k->new DLinkedList(node.freq)).add(node);
        
        while (freqMap.get(minFreq).isEmpty()) minFreq++;
        return node.value;
    }
    
    
    public int get(int key) {
        if(!keyMap.containsKey(key)) return -1;
        Node node = keyMap.get(key);
        return  _update(node);
        
    }
    
    public void put(int key, int value) {
        if(capacity==0) return;
        if (keyMap.containsKey(key)){
            Node node = keyMap.get(key);
            node.value = value;
            _update(node);
            return ;
        }
        if(size>=capacity){
            Node old = freqMap.get(minFreq).pop();
            keyMap.remove(old.key);
            size --;
        }
        Node node = new Node(key,value);
        freqMap.get(0).add(node);
        keyMap.put(key,node);
        minFreq = 0;
        size ++;
    }
}

File: 0460-lfu-cache/0460-lfu-cache.cpp
class DListNode{
public:
    int key;
    int value;
    int freq;
    DListNode *prev, *next;
    
    DListNode(int k, int v){
        key = k;
        value = v;
        freq = 1;
        prev = next = nullptr;
    }
};

class DList{
public:
    DListNode* sentinel;
    int size;
    
    DList(){
        size = 0;
        sentinel = new DListNode(-1, -1);
        sentinel->prev = sentinel->next = sentinel;
    }
    
    void prepend(DListNode* node){
        node->next = sentinel->next;
        node->prev = sentinel;
        sentinel->next->prev = node;
        sentinel->next = node;
        ++size;
    }
    
    DListNode* pop(DListNode* node = nullptr){
        if(size == 0) return nullptr;
        if(node == nullptr) node = sentinel->prev;
        node->prev->next = node->next;
        node->next->prev = node->prev;
        --size;
        return node;
    }
};

class LFUCache {
public:
    unordered_map<int, DListNode*> nodes;
    unordered_map<int, DList*> freqs;
    int capacity;
    int min_freq;
    
    LFUCache(int capacity) {
        this->capacity = capacity;
        min_freq = 0;
    }
    
    void _update(DListNode* node){
        if(freqs.find(node->freq) == freqs.end()){
            freqs[node->freq] = new DList();
        }
        if(freqs.find(node->freq+1) == freqs.end()){
            freqs[node->freq+1] = new DList();
        }
        freqs[node->freq]->pop(node);
        if(min_freq == node->freq && freqs[node->freq]->size == 0){
            ++min_freq;
        }
        
        ++node->freq;
        freqs[node->freq]->prepend(node);
    }
    
    int get(int key) {
        if(nodes.find(key) == nodes.end()){
            return -1;
        }
        
        DListNode* node = nodes[key];
        
        _update(node);
        
        return node->value;
    }
    
    void put(int key, int value) {
        if(capacity == 0) return;
        if(nodes.find(key) != nodes.end()){
            DListNode* node = nodes[key];
            _update(node);
            node->value = value;
        }else{
            if(nodes.size() == capacity){
                if(freqs.find(min_freq) == freqs.end()){
                    freqs[min_freq] = new DList();
                }
                
                DListNode* node = freqs[min_freq]->pop();
                if(node){
                    nodes.erase(node->key);
                }
            }
            
            DListNode* node = new DListNode(key, value);
            nodes[key] = node;
            if(freqs.find(1) == freqs.end()) freqs[1] = new DList();
            freqs[1]->prepend(node);
            min_freq = 1;
        }
    }
};